### ISSUE TITLE
data store: swap threads for processes

### ISSUE THREAD

**Issue by oliver-sanders** [cylc/cylc-uiserver#538]
We run subscriptions in a ThreadPoolExecutor.

In Python only one thread can actively run at a time (because of the GIL) so there is no compute parallelism advantage to this (but there may be IO concurrency advantages depending on the implementation details of the code being run).

This means that one large workflow can hog 100% of the CPU of the server, causing issues with other workflows.

We should be able to swap the ThreadPoolExecutor for a ProcessPoolExecutor. I had a quick try, but it didn't work first time so it's not a 30 second job.

This does raise the question of how many processes the UI Server should be allowed to spawn. I think we should be able to run more subscribers than processes in the pool, but would need to read the docs to find out.

**Comment by oliver-sanders** [comment]
https://github.com/cylc/cylc-uiserver/pull/548 may reduce the urgency on this by optimizing some resolver stuff.

**Comment by dwsutherland** [comment]
Yeah, reason I didn't go with separate processes is so the resolvers have access to the central data-store (if it works that way)

**Comment by dwsutherland** [comment]
Silly idea, and wider scope..
But if cylc-flow `DataStoreMgr` is refactored into `DeltaMgr` and a `DataStoreMgr` (which is imported by the UIS)..
This would mean UIS window size(s) are isolated from the Scheduler window size(s), so:
- Users cannot impact the scheduler machines with absurd window sizes.
- Different users/UIS viewing the same workflow can't effect each others window size

Would we even consider building a data-store for each subscription, advantages being:
- Different subscriptions can build different size views (including history).
- Each subscription can be it's own process (?), though, at the moment I'm not sure how this yields data back to the UIS.

First part, yes (and I think we've discussed it).. Second part, maybe?

**Comment by hjoliver** [comment]
Yes we originally wanted n=0 or 1 only, at the scheduler.

**Comment by oliver-sanders** [comment]
(the only reason we would want n=1 at the scheduler is Tui, we can drop back to n=0 if there isn't a client connected)

**Comment by oliver-sanders** [comment]
See also #194
